File: doodads/VB.NET/MyGeneration.dOOdads/DbAdapters/VisualFoxProEntity.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Data.Common
Imports System.Data.OleDb
Imports System.Text

Namespace MyGeneration.dOOdads
    Public Class VisualFoxProEntity
        Inherits BusinessEntity
        Public Sub New()
        End Sub

        Friend Overloads Overrides Function CreateDynamicQuery(ByVal entity As BusinessEntity) As DynamicQuery
            Return New VisualFoxProDynamicQuery(entity)
        End Function

        Friend Overloads Overrides Function CreateIDataParameter(ByVal name As String, ByVal value As Object) As IDataParameter
            Dim p As New OleDbParameter()
            p.ParameterName = name
            p.Value = value
            Return p
        End Function

        Friend Overloads Overrides Function CreateIDataParameter() As IDataParameter
            Return New OleDbParameter()
        End Function

        Friend Overloads Overrides Function CreateIDbCommand() As IDbCommand
            Return New OleDbCommand()
        End Function

        Friend Overloads Overrides Function CreateIDbDataAdapter() As IDbDataAdapter
            Return New OleDbDataAdapter()
        End Function

        Friend Overloads Overrides Function CreateIDbConnection() As IDbConnection
            Return New OleDbConnection()
        End Function

        Friend Overloads Overrides Function ConvertIDbDataAdapter(ByVal dataAdapter As IDbDataAdapter) As DbDataAdapter
            Return TryCast(TryCast(dataAdapter, OleDbDataAdapter), DbDataAdapter)
        End Function

        ' Overloaded in the generated class 
        Public Overridable Function GetAutoKeyColumn() As String
            Return ""
        End Function

        ' Called just before the Save() is truly executed 
        Protected Overloads Overrides Sub HookupRowUpdateEvents(ByVal adapter As DbDataAdapter)
            ' We only bother hooking up the event if we have an AutoKey 
            If Me.GetAutoKeyColumn().Length > 0 Then
                Dim da As OleDbDataAdapter = TryCast(adapter, OleDbDataAdapter)
                AddHandler da.RowUpdated, AddressOf OnRowUpdated
            End If
        End Sub

        ' If it's an Insert we fetch the @@Identity value and stuff it in the proper column 
        Protected Sub OnRowUpdated(ByVal sender As Object, ByVal e As OleDbRowUpdatedEventArgs)
            Try
                If e.Status = UpdateStatus.[Continue] AndAlso e.StatementType = StatementType.Insert Then
                    Dim txMgr As TransactionMgr = TransactionMgr.ThreadTransactionMgr()

                    Dim cmd As New OleDbCommand("SELECT @@IDENTITY")

                    ' We make sure we enlist in the ongoing transaction, otherwise, we 
                    ' would most likely deadlock 
                    txMgr.Enlist(cmd, Me)
                    Dim o As Object = cmd.ExecuteScalar()
                    ' Get the Identity Value 
                    txMgr.DeEnlist(cmd, Me)

                    If o IsNot Nothing Then
                        e.Row(Me.GetAutoKeyColumn()) = o
                        e.Row.AcceptChanges()
                    End If
                End If
            Catch
            End Try
        End Sub

        Protected Overloads Overrides Function _LoadFromRawSql(ByVal rawSql As String, ByVal ParamArray parameters As Object()) As IDbCommand
            Dim i As Integer = 0
            Dim token As String = ""
            Dim sIndex As String = ""
            Dim param As String = ""

            Dim cmd As New OleDbCommand()

            For Each o As Object In parameters
                sIndex = i.ToString()
                token = "{"c + sIndex + "}"c
                param = "@p" & sIndex

                rawSql = rawSql.Replace(token, param)

                Dim p As New OleDbParameter(param, o)
                cmd.Parameters.Add(p)
                i += 1
            Next

            cmd.CommandText = rawSql
            Return cmd
        End Function
    End Class
End Namespace

File: doodads/CSharp/MyGeneration.dOOdads/DbAdapters/VisualFoxProEntity.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Data.OleDb;
using System.Text;

namespace MyGeneration.dOOdads
{
	public class VisualFoxProEntity : BusinessEntity
	{
		public VisualFoxProEntity() { }

		override internal DynamicQuery CreateDynamicQuery(BusinessEntity entity)
		{
			return new VisualFoxProDynamicQuery(entity);
		}

		override internal IDataParameter CreateIDataParameter(string name, object value)
		{
			OleDbParameter p = new OleDbParameter();
			p.ParameterName = name;
			p.Value = value;
			return p;
		}

		override internal IDataParameter CreateIDataParameter()
		{
			return new OleDbParameter();
		}

		override internal IDbCommand CreateIDbCommand()
		{
			return new OleDbCommand();
		}

		override internal IDbDataAdapter CreateIDbDataAdapter()
		{
			return new OleDbDataAdapter();
		}

		override internal IDbConnection CreateIDbConnection()
		{
			return new OleDbConnection();
		}

		override internal DbDataAdapter ConvertIDbDataAdapter(IDbDataAdapter dataAdapter)
		{
			return (dataAdapter as OleDbDataAdapter) as DbDataAdapter;
		}

		// Overloaded in the generated class
		public virtual string GetAutoKeyColumn()
		{
			return "";
		}

		// Called just before the Save() is truly executed
		override protected void HookupRowUpdateEvents(DbDataAdapter adapter)
		{
			// We only bother hooking up the event if we have an AutoKey
			if (this.GetAutoKeyColumn().Length > 0)
			{
				OleDbDataAdapter da = adapter as OleDbDataAdapter;
				da.RowUpdated += new OleDbRowUpdatedEventHandler(OnRowUpdated);
			}
		}

		// If it's an Insert we fetch the @@Identity value and stuff it in the proper column
		protected void OnRowUpdated(object sender, OleDbRowUpdatedEventArgs e)
		{
			try
			{
				if (e.Status == UpdateStatus.Continue && e.StatementType == StatementType.Insert)
				{
					TransactionMgr txMgr = TransactionMgr.ThreadTransactionMgr();

					OleDbCommand cmd = new OleDbCommand("SELECT @@IDENTITY");

					// We make sure we enlist in the ongoing transaction, otherwise, we 
					// would most likely deadlock
					txMgr.Enlist(cmd, this);
					object o = cmd.ExecuteScalar(); // Get the Identity Value
					txMgr.DeEnlist(cmd, this);

					if (o != null)
					{
						e.Row[this.GetAutoKeyColumn()] = o;
						e.Row.AcceptChanges();
					}
				}
			}
			catch { }
		}

		override internal IDbCommand _LoadFromRawSql(string rawSql, params object[] parameters)
		{
			int i = 0;
			string token = "";
			string sIndex = "";
			string param = "";

			OleDbCommand cmd = new OleDbCommand();

			foreach (object o in parameters)
			{
				sIndex = i.ToString();
				token = '{' + sIndex + '}';
				param = "@p" + sIndex;

				rawSql = rawSql.Replace(token, param);

				OleDbParameter p = new OleDbParameter(param, o);
				cmd.Parameters.Add(p);
				i++;
			}

			cmd.CommandText = rawSql;
			return cmd;
		}
	}
}